% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/melonnpan_train.R
\name{melonnpan.train}
\alias{melonnpan.train}
\title{Model-based Genomically Informed High-dimensional Predictor 
of Microbial Community Metabolite Profiles}
\usage{
melonnpan.train(
  metab,
  metag,
  output,
  alpha = seq(0.05, 0.95, 0.05),
  lambda.choice = "lambda.1se",
  nfolds = 10,
  correction = "fdr",
  method = "spearman",
  cores = 4,
  seed = 1234,
  cutoff = 0.05,
  verbose = TRUE,
  no.transform.metab = FALSE,
  no.transform.metag = FALSE,
  discard.poor.predictions = FALSE,
  plot = FALSE,
  outputString = c("MelonnPan_Training_Summary", "MelonnPan_Trained_Weights",
    "MelonnPan_Trained_Metabolites")
)
}
\arguments{
\item{metab}{Training data of metabolite relative abundances (normalized). 
Must have the exact same rows (subjects/samples) as metag.}

\item{metag}{Training data of metagenomics sequence features' relative abundances (normalized).
Must have the exact same rows (subjects/samples) as metab.}

\item{output}{The output folder to write results.}

\item{alpha}{Grid of alpha values between 0 and 1. Default is `seq(0.05, 0.95, 0.05)`.}

\item{lambda.choice}{Choice of optimal lambda ('lambda.min' or 'lambda.1se'). Default is 'lambda.1se'.}

\item{nfolds}{Number of folds for internal cross-validation. Default is 10.}

\item{correction}{Multiplicity adjustment method, same as 'p.adjust'. Default is 'fdr'.}

\item{method}{Method to correlate measured and predicted metabolites ('spearman' or 'pearson'). Default is 'spearman'.}

\item{cores}{Number of cores to use for parallel processing. Default is 4.}

\item{seed}{Specify the arbitrary seed value for reproducibility. Default is 1234.}

\item{cutoff}{Q-value threshold for significant prediction. Default is 0.05.}

\item{verbose}{Should detalied message be printed. Default is TRUE.}

\item{no.transform.metab}{Should arcsine square root transformation (AST) be turned off for `metab`? 
Default is FALSE. If FALSE, `metab` must be proportional data ranging from 0 to 1.}

\item{no.transform.metag}{Should rank-based inverse normal (RIN) transformation be turned off for `metag`? 
Default is FALSE.}

\item{discard.poor.predictions}{Should predictions with model size = 1 be discarded? Default is FALSE.}

\item{plot}{Should CV error as a function of lambda be plotted. Default is FALSE.}

\item{outputString}{Names of the three output files. Default is 
'MelonnPan_Training_Summary', 'MelonnPan_Trained_Weights', and 'MelonnPan_Trained_Metabolites'.}
}
\description{
Predict metabolites from microbial sequence features' abundances. 
Both measurements are expected to be normalized before using MelonnPan.
}
\keyword{elastic}
\keyword{metabolite}
\keyword{metabolomics}
\keyword{metagenomics,}
\keyword{microbiome,}
\keyword{net,}
\keyword{prediction,}
